kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/PulsarStorageProducerIdManagerImpl.java

```diff
@@ -0,0 +1,197 @@
+                .topic(topic)
+                .accessMode(ProducerAccessMode.WaitForExclusive)
+                .blockIfQueueFull(true)
+                .createAsync();
+        return producerHandle.thenCompose(opProducer -> {
+            // nobody can write now to the topic
+            // wait for local cache to be up-to-date
+            CompletableFuture<Long> dummy = ensureLatestData(true)
+                    .thenCompose((___) -> {
+                        long start = nextId.get();
+                        List<Long> block = generateBlock(start + blockSize, blockSize);
+                        final long nextAvailableId = start + blockSize;
+                        // write to Pulsar
+                        byte[] serialized = BigInteger.valueOf(nextAvailableId).toByteArray();
+                        CompletableFuture<Long>  res =  opProducer
+                                .newMessage()
+                                .key("") // always the same key, this way we can rely on compaction
+                                .value(serialized)
+                                .sendAsync()
+                                .thenApply((msgId) -> {
+                                    log.debug("{} written {} as {}", this, nextAvailableId, msgId);
+                                    nextId.set(nextAvailableId);
+                                    availableIdsLocally.addAll(block);
+                                    long result = availableIdsLocally.remove();
+                                    log.debug("Returning allocated id {} for {}, new range: {}-{}",
+                                            result, topic, start, nextAvailableId);
+                                    return result;
+                                });
+                        return res;
+                    });
+            // ensure that we release the exclusive producer in any case
+            dummy.whenComplete((___, err) -> {
+                opProducer.closeAsync();
+            });
+            return dummy;
+        });
+    }
+
+    private static List<Long> generateBlock(long start, int blockSize) {
+        return LongStream.range(start, start + blockSize)
+                .boxed()
+                .collect(Collectors.toList());
+    }
+
+    public PulsarStorageProducerIdManagerImpl(String topicName, PulsarClient pulsarClient, int blockSize) {
+        this.topic = topicName;
+        this.pulsarClient = pulsarClient;
+        this.blockSize = blockSize;
+    }
+
+    public PulsarStorageProducerIdManagerImpl(String topicName, PulsarClient pulsarClient) {
+        this(topicName, pulsarClient, BLOCK_SIZE);
+    }
+
+    @Override
+    public CompletableFuture<Void> initialize() {
+       return ensureLatestData(false);
+    }
+
+    @Override
+    public synchronized void shutdown() {
+        if (reader != null) {
+            reader.whenComplete((r, e) -> {
+                if (r != null) {
+                    r.closeAsync();
```

Comment:
It's better to print the error message if closeAsync failed.

Reply:
fixed